File: src/main/java/com/cyprias/chunkspawnerlimiter/CslCommand.java
package com.cyprias.chunkspawnerlimiter;

import co.aikar.commands.BaseCommand;
import co.aikar.commands.annotation.CommandAlias;
import co.aikar.commands.annotation.CommandPermission;
import co.aikar.commands.annotation.Description;
import co.aikar.commands.annotation.Subcommand;
import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;

@CommandAlias("csl")
public class CslCommand extends BaseCommand {

    @Subcommand("reload")
    @CommandAlias("cslreload")
    @CommandPermission("csl.reload")
    @Description("Reloads the config file.")
    public void onReload(final CommandSender sender){
        ChunkSpawnerLimiter.getInstance().reloadConfig();
        Config.reload();
        ChatUtil.tell(sender,Config.Messages.RELOADED_CONFIG);
    }


    @Subcommand("info")
    @CommandPermission("csl.info")
    @Description("Shows config info.")
    public void onInfo(final CommandSender sender) {
        ChatUtil.tell(sender, "&2&l-- ChunkSpawnerLimiter v%s --",ChunkSpawnerLimiter.getInstance().getDescription().getVersion());
        ChatUtil.tell(sender,"&2&l-- Properties --");
        ChatUtil.tell(sender,"Debug Message: %s", Config.Properties.DEBUG_MESSAGES);
        ChatUtil.tell(sender,"Check Chunk Load: %s", Config.Properties.CHECK_CHUNK_LOAD);
        ChatUtil.tell(sender,"Check Chunk Unload: %s", Config.Properties.CHECK_CHUNK_UNLOAD);
        ChatUtil.tell(sender,"Active Inspection: %s",Config.Properties.ACTIVE_INSPECTIONS);
        ChatUtil.tell(sender,"Watch Creature Spawns: %s",Config.Properties.WATCH_CREATURE_SPAWNS);
        ChatUtil.tell(sender,"Check Surrounding Chunks: %s", Config.Properties.CHECK_SURROUNDING_CHUNKS);
        ChatUtil.tell(sender,"Inspection Frequency: %d",Config.Properties.INSPECTION_FREQUENCY);
        ChatUtil.tell(sender,"Notify Players: %s",Config.Properties.NOTIFY_PLAYERS);
        ChatUtil.tell(sender,"Preserve Named Entities: %s",Config.Properties.PRESERVE_NAMED_ENTITIES);
        ChatUtil.tell(sender,"Ignore Metadata: %s",Config.Properties.IGNORE_METADATA.toString());
        ChatUtil.tell(sender,"&2&l-- Messages --");
        ChatUtil.tell(sender,"Reloaded Config: %s",Config.Messages.RELOADED_CONFIG);
        ChatUtil.tell(sender,"Removed Entities: %s",Config.Messages.REMOVED_ENTITIES);
    }
}

File: spigot/src/main/kotlin/com/cyprias/chunkspawnerlimiter/commands/CslCommand.kt
package com.cyprias.chunkspawnerlimiter.commands

import co.aikar.commands.BaseCommand
import co.aikar.commands.CommandHelp
import co.aikar.commands.annotation.CommandAlias
import co.aikar.commands.annotation.CommandPermission
import co.aikar.commands.annotation.Description
import co.aikar.commands.annotation.HelpCommand
import co.aikar.commands.annotation.Subcommand
import com.cyprias.chunkspawnerlimiter.ChunkSpawnerLimiter
import com.cyprias.chunkspawnerlimiter.messages.Command
import org.bukkit.command.CommandSender
import org.bukkit.configuration.ConfigurationSection
import com.cyprias.chunkspawnerlimiter.utils.message

@CommandAlias("csl")
class CslCommand(private val plugin: ChunkSpawnerLimiter) : BaseCommand() {
    private val config = plugin.cslConfig

    @Subcommand(Command.Reload.COMMAND)
    @CommandAlias(Command.Reload.ALIAS)
    @CommandPermission(Command.Reload.PERMISSION)
    @Description(Command.Reload.DESCRIPTION)
    fun onReload(sender: CommandSender) {
        plugin.reloadConfigs()
        config.reloadedConfig?.let { sender.message(it) }
    }


    @Subcommand(Command.Settings.COMMAND)
    @CommandAlias(Command.Settings.ALIAS)
    @CommandPermission(Command.Settings.PERMISSION)
    @Description(Command.Settings.DESCRIPTION)
    fun onSettings(sender: CommandSender) {
        //todo convert to $
        sender.message("&2&l-- ChunkSpawnerLimiter v%s --", plugin.description.version)
        sender.message("&2&l-- Properties --")
        sender.message("Debug Message: %s", config.debugMessages)
        sender.message("Check Chunk Load: %s", config.checkChunkLoad)
        sender.message("Check Chunk Unload: %s", config.checkChunkUnload)
        sender.message("Active Inspection: %s", config.activeInspections)
        sender.message("Watch Creature Spawns: %s", config.watchCreatureSpawns)
        sender.message("Check Surrounding Chunks: %s", config.checkSurroundingChunks)
        sender.message("Inspection Frequency: %d", config.inspectionFrequency)
        sender.message("Notify Players: %s", config.notifyPlayers)
        sender.message("Preserve Named Entities: %s", config.preserveNamedEntities)
        sender.message("Ignore Metadata: %s", config.ignoreMetadata.toString())
        sender.message("Excluded Worlds: %s", config.excludedWorlds)
        sender.message("&2&l-- Messages --")
        sender.message("Reloaded Config: %s", config.reloadedConfig)
        sender.message("Removed Entities: %s", config.removedEntities)
    }

    @Subcommand(Command.Info.COMMAND)
    @CommandAlias(Command.Info.ALIAS)
    @CommandPermission(Command.Info.PERMISSION)
    @Description(Command.Info.DESCRIPTION)
    fun onInfo(sender: CommandSender) {
        sender.message("&2&l-- ChunkSpawnerLimiter v%s --", plugin.description.version)
        sender.message("&2&l-- Reasons to cull on: --")
        sendConfigurationSection(sender, config.spawnReasons())
        sender.message("&2&l-- Entity Limits: --")
        sendConfigurationSection(sender, config.entityLimits())
    }

    @HelpCommand
    fun onHelp(help: CommandHelp) = help.showHelp()

    private fun sendConfigurationSection(sender: CommandSender, section: ConfigurationSection) {
        for ((key, value) in section.getValues(false)) {
            sender.message("%s: %s", key, value.toString())
        }
    }
}